src/backend/aspen/app/views/sample.py

```diff
@@ -447,6 +445,23 @@ def create_sample():
                     user.group.name,
                 ]
 
+            valid_location: Optional[Location] = (
+                g.db_session.query(Location)
+                .filter(Location.id == data["sample"].get("location_id", None))
+                .one_or_none()
+            )
+            if valid_location:
+                sample_args["location_id"] = valid_location.id
+                sample_args["region"] = RegionType(valid_location.region)
+                sample_args["country"] = valid_location.country
+                sample_args["division"] = valid_location.division
+                sample_args["location"] = valid_location.location or ""
```

Comment:
There's currently a bug -- well, it could be intended, but I don't think it is -- regarding what happens when a user uploads a non-location-specific "location" for a sample's metadata. This is a screenshot of a Samples entry in table from the rdev when the user selects `North America/USA/California` as the Collection Location during its metadata upload:
![image](https://user-images.githubusercontent.com/89553795/143951842-08396907-8e15-45ea-a6da-91816bd4479a.png)
The `---` is where the Current Location should be. I believe this line is the specific cause, but not the root issue. The cause is because, when submitting `North America/USA/California`, that has no `location` aspect (it's `division` level only), so when the Samples table pulls that sample, there is an empty string for the `location`, so that gets parsed as false-y, which triggers the table cell renderer in FE to show it as `---`. Previously, submitting "California" as the collection location actually set the samples table `location` column to `California`. But now since we're truly matching the correct GISAID location, the lowest level `location` has no value for a general division. We need some kind of fallback for cases like this.

More generally though, I think there's a root issue of conflicting sources of truth regarding location. While the `samples` table still has `region`, `country`, `division`, and `location`, the addition of now having a FK `location_id` means we can have conflict between the sample's location info and the canonical location info provided by looking it up in the `locations` table. I don't have an easy answer for how to fix this -- the migration for legacy data could be hairy -- but I think we probably don't want two ways to look up the location.

I think we should either:
1. Not have a `location_id` for the samples. While the available options for location data (when user is submitting the metadata via dropdown, TSV, etc) do come from the `samples` table, they are just used as a way to populate the `samples` table. **Pro:** Easily backwards compatible with our legacy approach; **Con:** If the GISAID canonical names for locations drift, this could cause us pain. Also adds a bit more data storage overhead since we're duplicating values across many samples that could be consolidated.
2. Not have any of `region`, `country`, `division`, and `location` in the samples table, only have a `location_id` that is a FK to pulling the canonical values from that table. **Pro:** Normalized data, keeps us using canonical terms from GISAID; **Con:** Lots more code updates required, probably a reasonably painful DB migration to move old sample locations to new format, we need to watch GISAID locations for drift -- I've only skimmed the PR, don't really know how we ingest the GISAID locations, but if we're using their IDs for locations, we have to make sure that never changes under our feet or it could screw up all our locations so far.

I think (2) is the more "right" way to do it, but it's definitely more of a hassle. Also ties us more heavily to GISAID than we might want long-term if we're ever needing to provide a different set of locations for a non-COVID pathogen in the future. But as it stands, I'm not really sure what should happen if there's ever a conflict between the location info in a sample row and the `locations` row referenced by `location_id`.

Reply:
@vincent-czi just chiming in here -- #2 is where we're going, but we still have a fair bit of code that depends on the country/division/location/etc fields in the samples table, and this PR is already massive.

Our YAML builder project will require us to touch most of the code that works with sample locations, so the plan is to deprecate these old fields as part of that project.

Reply:
Okay, as long as there's a clear direction and it will be happening in the next couple months I think that's fine.

I do think we probably want some kind of helper func that can take either a Locations record or a dict of region, country, division, and location, then just return the first existing field of the most specific kind. That way we can solve for the "California" bug above.